feat(core:database): Migrate database module to KMP

HekmatullahAmin opened the pull request:
Fixes - [Jira-MM-112](https://mifosforge.jira.com/browse/MM-112)

Didn't create a Jira ticket, click [here](https://mifosforge.jira.com/jira/software/c/projects/MM/issues/) to create new.

Please Add Screenshots If there are any UI changes.

| Before                                     | After                                  |
|--------------------------------------------|----------------------------------------|
|                                            |                                        |


Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Run the static analysis check `./gradlew check` or `ci-prepush.sh` to make sure you didn't break anything

- [ ] If you have multiple commits please combine them into one commit by squashing them.

HekmatullahAmin committed:
feat(core:database): Migrate database module to KMP

niyajali committed:
fix: Typo in MifosNotificationDao

HekmatullahAmin committed:
Remove unnecessary files and codes

HekmatullahAmin committed:
Rename files and remove AppDatabaseConstructor from desktop

niyajali commented on core/database/build.gradle.kts:
```diff
@@ -8,22 +23,71 @@
-    alias(libs.plugins.mifos.android.hilt)
+    alias(libs.plugins.mifos.kmp.library)
     alias(libs.plugins.kotlin.serialization)
+    alias(libs.plugins.room)
```
Update these libs.plugins.mifos.android.room room convention plugin to KMP and use it here

HekmatullahAmin replied:
done

niyajali commented on core/database/build.gradle.kts:
```diff
@@ -8,22 +23,71 @@
         consumerProguardFiles("consumer-rules.pro")
     }
 
+    buildTypes {
```
remove this block's, not needed to here

HekmatullahAmin replied:
done

niyajali commented on core/database/build.gradle.kts:
```diff
@@ -8,22 +23,71 @@
+kotlin {
+
+    room {
+        schemaDirectory("$projectDir/schemas")
```
use convention plugin instead, and this block and the below platform dependencies could be moved to the plugin declaration..

HekmatullahAmin replied:
done

niyajali commented on core/database/src/androidMain/kotlin/org/mifos/mobile/core/database/di/DatabaseModule.android.kt:
```diff
@@ -0,0 +1,25 @@
+    single {
+        AppDatabaseFactory(androidApplication())
+            .createDatabase()
+            .setDriver(BundledSQLiteDriver())
```
add these configurations

      .fallbackToDestructiveMigrationOnDowngrade()
      .setDriver(BundledSQLiteDriver())
      .setQueryCoroutineContext(Dispatchers.IO) // Use MifosDispatcher.IO
      
      For more info - https://funkymuse.dev/posts/kmp-room-setup/

HekmatullahAmin replied:
I can use Dispatchers.IO but not MifosDispatchers.IO. Does core/common is implemented correctly? 
It gives a Type Mismatch warning.
<img width="1023" alt="Screenshot 2025-01-23 at 16 11 35" src="https://github.com/user-attachments/assets/09f4ebd6-6e35-45bf-b647-eef4ddfa35dd" />

HekmatullahAmin replied:
I already added the core.common dependency on the commonMain of database gradle

HekmatullahAmin replied:
added in this way ".setQueryCoroutineContext(get(named(MifosDispatchers.IO.name)))" it will fix the issue, as long as DispatchersModule is already included in the DI container

HekmatullahAmin committed:
refactor: Update AndroidRoomConventionPlugin to KMPRoomConventionPlugin

niyajali reviewed (approved):
@HekmatullahAmin We're merging this PR, And could you provide remaining module implementation on another PR using SQLDelight like for web JS & WasmJS

niyajali commented:
@HekmatullahAmin You haven't added Jira link in this PR description, move your ticket to in-progress or done and add link in the description.

HekmatullahAmin commented:
> @HekmatullahAmin You haven't added Jira link in this PR description, move your ticket to in-progress or done and add link in the description.

I’ve added the Jira link to the PR and moved it to the "Done" column. Regarding the SQLDelight implementation for Web JS and WasmJS, I plan to tackle it next week, as I’ll be quite busy in the coming week.
